### PR TITLE
junit5 migration

### PR DESCRIPTION
**Pull request by vshneer**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -44,18 +44,19 @@ Additional daml dependencies are required by the library, but these dependencies
...
           .parties(ALICE, BOB, CHARLIE)
           .build();
 
-  @ClassRule public static ExternalResource sandboxClassRule = sandbox.getClassRule();
-  @Rule public ExternalResource sandboxRule = sandbox.getRule();
+    ...
 ```
-The ClassRule and the Rule are mandatory parts of each integration test. They take care of starting/stopping the Sandbox. Note that `Sandbox sandbox` needs a DAR file path. DAR compilation should be done via build scripts.
-Sandbox has two modes, restart mode, in which it is restarted after each test case, and reset mode (use the `useReset()` function in the builder) that is faster but *cannot be supplied with a market setup scenario*.
+The @ExtendWith(SandboxTestExtension.class) is mandatory of each integration test class. It takes care of starting/stopping the Sandbox. Note that `Sandbox sandbox` needs a DAR file path. DAR compilation should be done via build scripts.
```
Comment (gaborh-da) [inline README.md:59]: > mandatory for each integration test class

File: build.sbt [hunk 2]
```diff
@@ -39,8 +39,8 @@ libraryDependencies ++= Seq(
   commonsIO,
   guava,
   scalaz,
-  junit4,
-  junitInterface,
+  junit5,
+  "net.aichler" % "jupiter-interface" % JupiterKeys.jupiterVersion.value % Test,
```
Comment (gaborh-da) [inline build.sbt:43]: What is this?

File: src/test/java/com/daml/extensions/testing/PartiesAllocationTest.java [hunk 3]
```diff
@@ -28,8 +31,12 @@ public void testPartiesAreAllocated() {
     sandbox.getPartyId(CHARLIE);
   }
 
-  @Test(expected = NullPointerException.class)
+  @Test
   public void notAllocatedPartyThrows() throws NullPointerException {
-    sandbox.getPartyId("notAllocatedParty");
```
Comment (gaborh-da) [inline src/test/java/com/daml/extensions/testing/PartiesAllocationTest.java:33]: Did this nice annotation go away (null pointer exception)?

File: src/test/java/com/daml/extensions/testing/SetupAppCallbackWithRestartIT.java [hunk 4]
```diff
@@ -28,11 +29,12 @@ public class SetupAppCallbackWithRestartIT {
           .setupAppCallback(client -> boolFlag.set(true))
           .build();
 
-  @ClassRule public static ExternalResource sandboxClassRule = sandbox.getClassRule();
-  @Rule public ExternalResource sandboxRule = sandbox.getRule();
+  public Sandbox getSandbox() {
```
Comment (gaborh-da) [inline src/test/java/com/daml/extensions/testing/SetupAppCallbackWithRestartIT.java:32]: Why is this getter needed?

File: build.sbt (hunk 2 above)
Comment (vshneer) [reply build.sbt:43]: ah! thx!
junit5 - ok stands for junit5 dependency
the next line is updated version of the dependency that previously lived inside `junitInterface` variable
this "interface" give access to test trough sbt command line. So you can type `sbt test` or `sbt testOnly @SomeTest` this is crucial for our CI
I will wrap it back to the `junitInterface` var in a minute

File: src/test/java/com/daml/extensions/testing/SetupAppCallbackWithRestartIT.java (hunk 4 above)
Comment (vshneer) [reply src/test/java/com/daml/extensions/testing/SetupAppCallbackWithRestartIT.java:32]: we need to restart sandbox after each test inside test class
actually this is an action on SandboxManager object that lives inside Sandbox object
this object is private 
before, with junit4 we used methods that lived in the Sandbox class
and we call them from @Rull and @ClassRule properties
these properties are deleted in junit5
the new approach is to create a separate class that implements these AfterEach methods
these methods takes Test context as a parameter
so what I needed to do is to extract sanboxManager from this context
you can see this code in the new file [SandboxTestExtension.java](https://github.com/digital-asset/lib-daml-jvm-test/pull/81/files#diff-3cca91a6e9d34585b84258325f314adb66d754dade767a7420770e9e61456834)
```
    Object testInstance = context.getRequiredTestInstance();
    Sandbox sandbox =
        (Sandbox) testInstance.getClass().getMethod("getSandbox", null).invoke(testInstance, null);
    sandbox.stop();
```
I can't find a way to extract sandboxManager from context without this getter
May be there is a better approach in java for it... The thing I don't like is that I add this getter in every test class so it is badly duplicated
I tried to add a super class with sanbox protected and this method
but this mean that every test should override sandbox property.
As I learn yesterday overriding properties is not a good idea in java

Comment (gaborh-da) [reply src/test/java/com/daml/extensions/testing/SetupAppCallbackWithRestartIT.java:32]: is it not possible to use the property itself (without the getter)?

Comment (gaborh-da) [reply src/test/java/com/daml/extensions/testing/SetupAppCallbackWithRestartIT.java:32]: https://docs.oracle.com/javase/8/docs/api/java/lang/Class.html#getField-java.lang.String-

File: src/test/java/com/daml/extensions/testing/PartiesAllocationTest.java (hunk 3 above)
Comment (vshneer) [reply src/test/java/com/daml/extensions/testing/PartiesAllocationTest.java:33]: https://www.baeldung.com/junit-5-migration#1-annotations

yep...

File: src/main/java/com/daml/extensions/testing/junit5/SandboxTestExtension.java [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+        .ifPresent(
+            testInstance -> {
+              List<Field> sandboxFields =
+                  AnnotationSupport.findAnnotatedFields(testInstance.getClass(), TestSandbox.class);
+              for (Field sandboxField : sandboxFields) {
+                try {
+                  Sandbox sandbox = (Sandbox) sandboxField.get(testInstance);
+                  sandbox.stop();
+                } catch (IllegalAccessException e) {
+                  e.printStackTrace();
```
Comment (gaborh-da) [inline src/main/java/com/daml/extensions/testing/junit5/SandboxTestExtension.java:34]: Shouldn't we just rather throw some runtime (unchecked) exception?
I am not sure printing stack trace and continuing would make anyone's life easier.

File: src/test/java/com/daml/extensions/testing/junit5/SandboxIT.java [hunk 6]
```diff
@@ -40,9 +39,4 @@ public void specifiedPortIsAssignedWhenSandboxIsStarted() {
   public void ledgerIdSpecified() {
     assertThat(sandbox.getLedgerId(), is("sample-ledger"));
   }
-
-  private void assertsIsBetween(int x, int low, int high) {
```
Comment (gaborh-da) [inline src/test/java/com/daml/extensions/testing/junit5/SandboxIT.java:44]: @vshneer why is this deleted?

Comment (vshneer) [reply src/test/java/com/daml/extensions/testing/junit5/SandboxIT.java:44]: there was a test that checked if auto assigned port is between a range boundaries. 
I put a new test instead of it - check if custom port assigned. 
Because I cannot test these two in one test class - sandbox port either custom or auto assigned...
Maybe I should return this boundary test back, and for custom test create another class. I'll do it

Comment (gaborh-da) [reply src/test/java/com/daml/extensions/testing/junit5/SandboxIT.java:44]: I would advise having two separate tests (classes) to test them, then.

**Review by gaborh-da — APPROVED** [review]
